perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AuthzResolverBlImpl.java

```diff
@@ -1178,10 +1137,56 @@ public static void unsetRole(PerunSession sess, Group authorizedGroup, String ro
 	 * @param role                role of user in a session ( PERUNADMIN | VOADMIN | GROUPADMIN | SELF | FACILITYADMIN | VOOBSERVER | TOPGROUPCREATOR | RESOURCESELFSERVICE | RESOURCEADMIN )
 	 * @param complementaryObject object for which role will be unset
 	 */
-	public static void unsetRole(PerunSession sess, Group authorizedGroup, PerunBean complementaryObject, String role) throws GroupNotAdminException {
-		List<PerunBean> complementaryObjects = new ArrayList<>();
-		complementaryObjects.add(complementaryObject);
-		AuthzResolverBlImpl.unsetRole(sess, authorizedGroup, role, complementaryObjects);
+	public static void unsetRole(PerunSession sess, Group authorizedGroup, PerunBean complementaryObject, String role) throws GroupNotAdminException, RoleCannotBeManagedException {
+		if (!objectAndRoleManageableByEntity(authorizedGroup, complementaryObject, role)) {
+			throw new RoleCannotBeManagedException(role, complementaryObject, authorizedGroup);
+		}
+
+		Map<String, Integer> mappingOfValues = createMappingOfValues(authorizedGroup, complementaryObject, role);
+
+		try {
+			authzResolverImpl.unsetRole(sess, mappingOfValues, role);
+		} catch (RoleNotSetException e) {
+			throw new GroupNotAdminException("Group id=" + authorizedGroup.getId() + " is not "+role+" in " + complementaryObject, e);
+		}
+
+		getPerunBl().getAuditer().log(sess, new RoleUnsetForGroup(complementaryObject, authorizedGroup, role));
+
+		if (role.equals(Role.SPONSOR) && complementaryObject.getBeanName().equals("Vo"))
+			getPerunBl().getVosManagerBl().handleGroupLostVoRole(sess, authorizedGroup, (Vo) complementaryObject, Role.SPONSOR);
+
+		if (authorizedGroup != null && sess.getPerunPrincipal() != null && sess.getPerunPrincipal().getUser() != null) {
+			List<Member> groupMembers = perunBl.getGroupsManagerBl().getGroupMembers(sess, authorizedGroup);
+			List<Member> userMembers = perunBl.getMembersManagerBl().getMembersByUser(sess, sess.getPerunPrincipal().getUser());
+			userMembers.retainAll(groupMembers);
+			if (!userMembers.isEmpty()) AuthzResolverBlImpl.refreshAuthz(sess);
+		}
+	}
+
+	/**
+	 * Add owner for a specific user.
+	 *
+	 * @param sess Principal's session
+	 * @param user - owner
+	 * @param complementaryObject specific user
+	 * @throws AlreadyAdminException
+	 */
+	public static void addSpecificUserOwner(PerunSession sess, User user, PerunBean complementaryObject) throws AlreadyAdminException {
```

Comment:
Is there any reason to use a complementary object when you are always recasting it to the User object immediately?

Reply:
Same for removeSpecificUserOwner method.

Reply:
No, copy-paste mistake probably. Fixed.